File: Product of Array Except Self - Leetcode 238/Product of Array Except Self - Leetcode 238.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> result(n);
        
        int leftProduct = 1;
        int rightProduct = 1;
        
        // Calculate left products
        for (int i = 0; i < n; i++) {
            result[i] = leftProduct;
            leftProduct *= nums[i];
        }
        
        // Calculate right products and final result
        for (int i = n - 1; i >= 0; i--) {
            result[i] *= rightProduct;
            rightProduct *= nums[i];
        }
        
        return result;
    }
};

File: Product of Array Except Self - Leetcode 238/Product of Array Except Self - Leetcode 238.go
package solution

func productExceptSelf(nums []int) []int {
    res := make([]int , len(nums))
    
    res[0] = 1
    
    for index := 1 ; index < len(nums) ; index++ {
        res[index] = nums[index - 1] * res[index - 1];
    }
    
    suffix := 1;
    
    for index := len(nums) - 1 ; index >= 0 ;index-- {
        res[index] *= suffix
        suffix *= nums[index]
    }
    return res
}
